File: app/src/main/java/org/progresspics/android/LayoutPicker.java
package org.progresspics.android;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

public class LayoutPicker {

    public interface Listener {
        void onItemClicked(int[] cellsPerRow);
    }

    @NonNull
    public static View inflate(Context context, Listener listener) {
        View content = LayoutInflater.from(context).inflate(R.layout.layout_picker, null);

        RecyclerView recyclerView = content.findViewById(R.id.layout_list);
        recyclerView.setHasFixedSize(true);

        RecyclerView.LayoutManager layoutManager = new GridLayoutManager(context, 4);
        recyclerView.setLayoutManager(layoutManager);

        RecyclerView.Adapter adapter = new Adapter(
                listener,
                context.getResources().getStringArray(R.array.layouts_array));
        recyclerView.setAdapter(adapter);

        return content;
    }

    private static class Adapter extends RecyclerView.Adapter {
        private Listener mListener;
        private final String[] mDescriptors;

        Adapter(Listener listener, String[] descriptors) {
            mListener = listener;
            mDescriptors = descriptors;
        }

        @NonNull
        @Override
        public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            View itemView = inflater.inflate(R.layout.layout_sample_grid, null);
            return new ViewHolder(itemView);
        }

        @Override
        public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
            final int [] cellsPerRow = parseLayoutDescriptor(mDescriptors[position]);

            View itemView = ((ViewHolder) holder).mItemView;
            itemView.setOnClickListener((v) -> mListener.onItemClicked(cellsPerRow));

            ViewGroup grid = (ViewGroup) ((ViewGroup) itemView).getChildAt(0);

            for (int r = 0; r != grid.getChildCount(); r++) {
                final boolean activeRow = r < cellsPerRow.length;
                final ViewGroup row = (ViewGroup) grid.getChildAt(r);
                row.setVisibility(activeRow ? View.VISIBLE : View.GONE);
                for (int c = 0; c != row.getChildCount(); c++) {
                    final boolean activeColumn = activeRow && c < cellsPerRow[r];
                    final View cell = row.getChildAt(c);
                    cell.setVisibility(activeColumn ? View.VISIBLE : View.GONE);
                }
            }
        }

        @Override
        public int getItemCount() {
            return mDescriptors.length;
        }
    }

    private static class ViewHolder extends RecyclerView.ViewHolder {
        View mItemView;
        ViewHolder(View grid) {
            super(grid);
            mItemView = grid;
        }
    }

    private static int[] parseLayoutDescriptor(String descriptor) {
        String[] parts = descriptor.split("x");
        if (parts.length == 2) {
            int rows = Integer.parseInt(parts[1]);
            int columns = Integer.parseInt(parts[0]);
            int [] cellsPerRow = new int[rows];
            for (int r = 0; r != rows; ++r) {
                cellsPerRow[r] = columns;
            }
            return cellsPerRow;
        }
        parts = descriptor.split("\\+");
        if (parts.length > 1) {
            int [] cellsPerRow = new int[parts.length];
            for (int r = 0; r != parts.length; ++r) {
                cellsPerRow[r] = Integer.parseInt(parts[r]);
            }
            return cellsPerRow;
        }
        return null;
    }
}

File: app/src/main/java/org/progresspics/android/LayoutPicker.kt
package org.progresspics.android

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Resources
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import java.util.Arrays

object LayoutPicker {

    interface Listener {
        fun onItemClicked(cellsPerRow: IntArray)
    }

    private fun getDescriptors(resources: Resources): Array<String> {
        return resources.getStringArray(R.array.layouts_array)
    }

    fun findBestLayout(resources: Resources, count: Int): IntArray {
        val descriptors = getDescriptors(resources)
        for (descriptor in descriptors) {
            val layout = parseLayoutDescriptor(descriptor)
            val layoutSize = layout.fold(0) { sum, value -> sum + value }
            if (layoutSize >= count) {
                return layout
            }
        }
        return parseLayoutDescriptor(descriptors[descriptors.size - 1])
    }

    @SuppressLint("InflateParams")
    fun inflate(context: Context, listener: Listener): View {
        val content = LayoutInflater.from(context).inflate(R.layout.layout_picker, null)

        val recyclerView = content.findViewById<RecyclerView>(R.id.layout_list)
        recyclerView.setHasFixedSize(true)
        recyclerView.layoutManager = GridLayoutManager(context, 4)
        recyclerView.adapter = Adapter(listener, getDescriptors(context.resources))

        return content
    }

    private class Adapter internal constructor(
            private val listener: Listener, private val descriptors: Array<String>)
        : RecyclerView.Adapter<ViewHolder>() {

        override fun getItemCount(): Int {
            return descriptors.size
        }

        @SuppressLint("InflateParams")
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(LayoutInflater.from(parent.context).inflate(
                    R.layout.layout_sample_grid, null))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val cellsPerRow = parseLayoutDescriptor(descriptors[position])

            holder.itemView.setOnClickListener { listener.onItemClicked(cellsPerRow) }

            val grid = (holder.itemView as ViewGroup).getChildAt(0) as ViewGroup

            for (r in 0 until grid.childCount) {
                val activeRow = r < cellsPerRow.size
                val row = grid.getChildAt(r) as ViewGroup
                row.visibility = if (activeRow) View.VISIBLE else View.GONE
                for (c in 0 until row.childCount) {
                    val activeColumn = activeRow && c < cellsPerRow[r]
                    val cell = row.getChildAt(c)
                    cell.visibility = if (activeColumn) View.VISIBLE else View.GONE
                }
            }
        }
    }

    private class ViewHolder internal constructor(itemView: View)
        : RecyclerView.ViewHolder(itemView)

    private fun parseLayoutDescriptor(descriptor: String): IntArray {
        val dimensions = descriptor.split("x".toRegex()).toTypedArray()
        if (dimensions.size == 2) {
            val rows = Integer.parseInt(dimensions[1])
            val columns = Integer.parseInt(dimensions[0])
            val cellsPerRow = IntArray(rows)
            Arrays.fill(cellsPerRow, columns)
            return cellsPerRow
        }
        val rowCounts = descriptor.split("\\+".toRegex()).toTypedArray()
        return rowCounts.map { Integer.parseInt(it) }.toIntArray()
    }
}